pkg/generic/descriptor/http.go

```diff
@@ -34,19 +36,115 @@ const (
 
 // HTTPRequest ...
 type HTTPRequest struct {
-	Header      http.Header
-	Query       url.Values
-	Cookies     Cookies
-	Method      string
-	Host        string
-	Path        string
 	Params      *Params // path params
+	Request     *http.Request
 	RawBody     []byte
 	Body        map[string]interface{}
 	GeneralBody interface{} // body of other representation, used with ContentType
 	ContentType MIMEType
 }
 
+type jsonCache struct {
+	root ast.Node
+	m    map[string]string
+}
+
+// GetHeader implements RequestGetter.GetHeader of dynamicgo
+func (req HTTPRequest) GetHeader(key string) string {
+	return req.Request.Header.Get(key)
+}
+
+// GetCookie implements RequestGetter.GetCookie of dynamicgo
+func (req HTTPRequest) GetCookie(key string) string {
+	if c, err := req.Request.Cookie(key); err == nil {
+		return c.Value
+	}
+	return ""
+}
+
+// GetQuery implements RequestGetter.GetQuery of dynamicgo
+func (req HTTPRequest) GetQuery(key string) string {
+	return req.Request.URL.Query().Get(key)
+}
+
+// GetBody implements RequestGetter.GetBody of dynamicgo
+func (req HTTPRequest) GetBody() []byte {
+	return req.RawBody
+}
+
+// GetMethod implements RequestGetter.GetMethod of dynamicgo
+func (req HTTPRequest) GetMethod() string {
+	return req.Request.Method
+}
+
+// GetPath implements RequestGetter.GetPath of dynamicgo
+func (req HTTPRequest) GetPath() string {
+	return req.Request.URL.Path
+}
+
+// GetHost implements RequestGetter.GetHost of dynamicgo
+func (req HTTPRequest) GetHost() string {
+	return req.Request.Host
+}
+
+// GetParam implements RequestGetter.GetParam of dynamicgo
+func (req HTTPRequest) GetParam(key string) string {
+	return req.Params.ByName(key)
+}
+
+// GetMapBody implements RequestGetter.GetMapBody of dynamicgo
+func (req HTTPRequest) GetMapBody(key string) string {
+	if len(req.RawBody) == 0 {
+		return ""
+	}
+	if req.GeneralBody == nil && req.Request != nil {
+		body := req.RawBody
+		s := utils.SliceByteToString(body)
+		node := ast.NewRaw(s)
+		cache := &jsonCache{
+			root: node,
+			m:    make(map[string]string),
+		}
+		req.GeneralBody = cache
+	}
+	switch t := req.GeneralBody.(type) {
+	case *jsonCache:
+		// fast path
+		if v, ok := t.m[key]; ok {
```

Comment:
Why do you design the `fast path`? Is it possible to access the same key repeatedly?

Reply:
@jayantxie It is possible in dynamicgo in theory.

Reply:
U can try to remove this `jsonCache`, if there is little repeat key in annotation in practice it will be slower instead

Reply:
done